src/ballet/txn/fd_txn_parse.c

```diff
@@ -184,39 +194,120 @@ fd_txn_parse( uchar const             * payload,
-    for( ulong k=0; k<parsed->instr[ j ].acct_cnt; k++ ) {
-      CHECK( payload[ parsed->instr[ j ].acct_off + k ] < acct_addr_cnt + addr_table_adtl_cnt );
+  if (NULL != parsed) {
+    /* Final validation that all the account address indices are in range */
+    for( ulong j=0; j<instr_cnt; j++ ) {
+      for( ulong k=0; k<parsed->instr[ j ].acct_cnt; k++ ) {
+        CHECK( payload[ parsed->instr[ j ].acct_off + k ] < acct_addr_cnt + addr_table_adtl_cnt );
+      }
     }
+
+    /* Assign final variables */
+    parsed->addr_table_lookup_cnt         = (uchar)addr_table_cnt;
+    parsed->addr_table_adtl_writable_cnt  = (uchar)addr_table_adtl_writable_cnt;
+    parsed->addr_table_adtl_cnt           = (uchar)addr_table_adtl_cnt;
+    parsed->_padding_reserved_1           = (uchar)0;
   }
-  /* Assign final variables */
-  parsed->addr_table_lookup_cnt         = (uchar)addr_table_cnt;
-  parsed->addr_table_adtl_writable_cnt  = (uchar)addr_table_adtl_writable_cnt;
-  parsed->addr_table_adtl_cnt           = (uchar)addr_table_adtl_cnt;
-  parsed->_padding_reserved_1           = (uchar)0;
 
-  if( FD_LIKELY( counters_opt ) ) counters_opt->success_cnt++;
+  if( FD_LIKELY( counters_opt   ) ) counters_opt->success_cnt++;
+  if( FD_LIKELY( payload_sz_opt ) ) *payload_sz_opt = i;
   return fd_txn_footprint( instr_cnt, addr_table_cnt );
 
   #undef CHECK
   #undef CHECK_LEFT
   #undef READ_CHECKED_COMPACT_U16
 }
+
+ulong
+fd_txn_xray( uchar const             * payload,
+             ulong                     payload_sz,
+             fd_txn_xray_result_t    * result ) {
+  ulong i = 0UL;
+
+#define READ_COMPACT_U16( var_name, where )                              \
+  do {                                                                   \
+    ulong _where = (where);                                              \
+    ulong _out_sz = fd_cu16_dec_sz( payload+_where, payload_sz-_where ); \
+    if ( FD_UNLIKELY( !_out_sz ) ) return 0UL;                           \
+    (var_name) = fd_cu16_dec_fixed( payload+_where, _out_sz );           \
+    (where) += _out_sz;                                                  \
+  } while( 0 )
+
+  if ( FD_UNLIKELY( i >= payload_sz ) ) return 0UL;
+  uchar signature_cnt  = payload[ i ]; i++;
+  ulong signature_off  =          i  ;
+  i += FD_TXN_SIGNATURE_SZ*signature_cnt;
+  result->signature_cnt = signature_cnt;
+  result->signature_off = (ushort)signature_off;
+
+  if ( FD_UNLIKELY( i >= payload_sz ) ) return 0UL;
+  uchar header_b0      = payload[ i ]; i++;
+  uchar transaction_version;
+  if( FD_LIKELY( (ulong)header_b0 & 0x80UL ) ) {
+    /* This is a versioned transaction */
+    transaction_version = header_b0 & 0x7F;
+    /* Only recognized one so far */
+    i++;
+  } else {
+    transaction_version = FD_TXN_VLEGACY;
+  }
+  i++;
+  i++;
+
+  ushort acct_addr_cnt = (ushort)0;
+  READ_COMPACT_U16( acct_addr_cnt, i );
+  i += FD_TXN_ACCT_ADDR_SZ*acct_addr_cnt;
+  i += FD_TXN_BLOCKHASH_SZ;
```

Comment:
Unless this is designed for only trusted inputs, this function is missing a bunch of bounds checks. If acct_addr_cnt is huge, then `i` will go past payload_sz. `payload_sz-_where` will underflow.  It'll read a bunch of data past the end of the array, and I'm not convinced the final check at the end will necessarily catch it, because it could maybe overflow again.

Reply:
added in bounds checks

Reply:
I don't think they're quite right. E.g. the first check checks `payload_sz>=0` when it needs to check `payload_sz>=1`. The one on line 261 doesn't protect the read that follows of the header, etc.

Reply:
how is this?